gokitmiddlewares/timeoutmiddleware/middleware.go

```diff
@@ -0,0 +1,76 @@
+package timeoutmiddleware
+
+import (
+	"context"
+
+	"github.com/arquivei/foundationkit/errors"
+	"github.com/go-kit/kit/endpoint"
+)
+
+// New returns a new timeout middleware.
+//
+// After timeout is reached, if the middleware is configured to wait,
+// it will just cancel the context and wait for next endpoint to return.
+// But if the middleware is configured to not wait, it will run the next endpoint
+// inside a go-routine and return error as soon as the context is canceled.
+func New(c Config) (endpoint.Middleware, error) {
+	return func(next endpoint.Endpoint) endpoint.Endpoint {
+		// Timeout is disabled
+		if c.Timeout <= 0 {
+			return next
+		}
+
+		return func(ctx context.Context, request interface{}) (response interface{}, err error) {
+			ctx, cancel := context.WithTimeout(ctx, c.Timeout)
+			defer cancel()
+
+			defer func() {
+				if err != nil && ctx.Err() != nil {
+					err = errors.E(err, c.ErrorSeverity)
+				}
+			}()
+
+			if c.Wait {
+				return next(ctx, request)
+			}
+
+			select {
```

Comment:
Nit: this is a situation in which I think that using an `else` on `c.Wait` would increase readability

Reply:
I think else is not idiomatic because of the return, but I extracted the logic to a new function.